test/manifolds/segre.jl

```diff
@@ -0,0 +1,4149 @@
+
+                # Control points
+                y1 = y[1:n]
+                y2 = y[(n + 1):(2 * n)]
+                y3 = y[(2 * n + 1):(3 * n)]
+                y4 = y[(3 * n + 1):(4 * n)]
+
+                # Bezier curve from 0 to v with control points y1, ..., y4
+                function b(t)
+                    return (
+                        (1 - t)^5 * x0 +
+                        5 * t * (1 - t)^4 * (x1 + y1) +
+                        10 * t^2 * (1 - t)^3 * (x2 + y2) +
+                        10 * t^3 * (1 - t)^2 * (x3 + y3) +
+                        5 * t^4 * (1 - t) * (x4 + y4) +
+                        t^5 * x5
+                    )
+                end
+
+                # Length of curve on manifold
+                ps_ = [exp(M, p, get_vector(M, p, b(t))) for t in 0.0:1e-3:1.0]
+                ds = [distance(M, p1, p2) for (p1, p2) in zip(ps_[1:(end - 1)], ps_[2:end])]
+                return sum(ds)
+            end
+
+            # dy = rand(4 * n); dy = dy / norm(dy)
+            f = a -> curve_length(a * dy)
+            @test(isapprox(finite_difference(f, 0.0, 1e-3), 0.0; atol=1e-5))
+            @test(finite_difference(f, 0.0, 1e-3; order=2) >= 0.0)
+        end
+
+        @testset "log()" begin
+            @test(is_vector(M, p, log(M, p, q)))
+            @test(isapprox(0.0 * v, log(M, p, p); atol=1e-5))
+            @test(isapprox(q, exp(M, p, log(M, p, q)), atol=1e-10))
+        end
+
+        @testset "riemann_tensor()" begin
+
+            # Test Riemann tensor by testing that sectional curvature is difference
+            # between circumference and 2 pi r for small circles. Since the sectional
+            # curvature contains the same information as the Riemann tensor, this
+            # should be fine.
+
+            sectional_curvature(M, p, u, v) =
+                inner(M, p, riemann_tensor(M, p, u, v, v), u) /
+                (inner(M, p, u, u) * inner(M, p, v, v) - inner(M, p, u, v)^2)
+
+            # Orthonormalize
+            u_ = u / norm(M, p, u)
+            v_ = v - inner(M, p, u_, v) * u_
+            v_ = v_ / norm(M, p, v_)
+
+            r = 1e-2
+            ps_ = [
+                exp(M, p, r * (cos(theta) * u_ + sin(theta) * v_)) for
+                theta in 0.0:1e-3:(2 * pi)
+            ]
+            ds = [distance(M, p1, p2) for (p1, p2) in zip(ps_, [ps_[2:end]..., ps_[1]])]
+            C = sum(ds)
+            K = 3 * (2 * pi * r - C) / (pi * r^3) # https://en.wikipedia.org/wiki/Bertrand%E2%80%93Diguet%E2%80%93Puiseux_theorem
+
+            @test(isapprox(K, sectional_curvature(M, p, u, v); rtol=1e-2, atol=1e-2))
+        end
+
+        @testset "get_coordinates()" begin
+            @test(isapprox(v, get_vector(M, p, get_coordinates(M, p, v))))
+        end
+
+        @testset "get_vector()" begin
+            @test(isapprox(X, get_coordinates(M, p, get_vector(M, p, X))))
+        end
+    end
+end
+
+# TODO: Test that distance and inner are compatible
```

Comment:
`sectional_curvature` seems to be the main thing missing (besides embedding and a few points/vectors that fail the checks on purpose).

Reply:
... for both the warped and non-warped metric, so doing that once in the for loop should be fine :)